Regression: Cannot send message: `SocketException: Connection reset`

### Checklist

- [X] I have used the search function to see if someone else has already submitted the same bug report.
- [X] I will describe the problem with as much detail as possible.

### App version

main branch (d93f690c0e016e324c44abe36f8cf8fa32a39114)

### Where did you get the app from?

Other

### Android version

Android 11 (LineageOS)

### Device model

OnePlus 3T

### Steps to reproduce

1. Compose E-Mail
2. Press "send" button

(see logs section for `git bisect`)

### Expected behavior

E-Mail is sent.

### Actual behavior

I get a notification saying `SocketException: Connection reset` and the message stays in the outbox.

### Logs

### Output from `git bisect`:

> There are only 'skip'ped commits left to test.
> The first bad commit could be any of:
> 2f258d6886f5c7d4d1130e0a3248f8757ef116e8
> 56da545493cd53bc89b5ddec4c2086ab53284710
> We cannot bisect more!

### Logcat:

```
MessagingController: Running command 'sendPendingMessages', seq = 3 (background priority)
LocalMessage: No headers available for this message!
LocalMessage: No headers available for this message!
MessagingController: Scanning Outbox folder for messages to send
MessagingController: Send count for message K9LOCAL:1cca1e67-5b6d-4310-86d9-5c828adb4f75 is 0
LockableDatabase: LockableDatabase: Transaction ended, took 25 ms / com.fsck.k9.mailstore.LocalMessage.setFlag(LocalMessage.java:260)
MessagingController: Sending message with UID K9LOCAL:1cca1e67-5b6d-4310-86d9-5c828adb4f75
SmtpTransport: SMTP >>> QUIT
SmtpTransport: SMTP <<< 220 redacted.com ESMTP Postfix (Debian/GNU)
SmtpTransport: SMTP >>> EHLO [127.0.0.1]
CompatibilityChangeReporter: Compat change id reported: 150939131; UID 10512; state: ENABLED
SmtpTransport: SMTP <<< 250-redacted.com
SmtpTransport: SMTP <<< 250-PIPELINING
SmtpTransport: SMTP <<< 250-SIZE 52428800
SmtpTransport: SMTP <<< 250-ETRN
SmtpTransport: SMTP <<< 250-STARTTLS
SmtpTransport: SMTP <<< 250-ENHANCEDSTATUSCODES
SmtpTransport: SMTP <<< 250-8BITMIME
SmtpTransport: SMTP <<< 250-DSN
SmtpTransport: SMTP <<< 250 CHUNKING
SmtpTransport: SMTP >>> STARTTLS
SmtpTransport: SMTP <<< 220 2.0.0 Ready to start TLS
DefaultTrustedSocketFactory: Could not call SSLSocket#setHostname(String) method 
DefaultTrustedSocketFactory: java.lang.NoSuchMethodException: com.android.org.conscrypt.Java8EngineSocket.setHostname [class java.lang.String]
DefaultTrustedSocketFactory: 	at java.lang.Class.getMethod(Class.java:2072)
DefaultTrustedSocketFactory: 	at java.lang.Class.getMethod(Class.java:1693)
DefaultTrustedSocketFactory: 	at com.fsck.k9.mail.ssl.DefaultTrustedSocketFactory.setHostnameViaReflection(DefaultTrustedSocketFactory.java:158)
DefaultTrustedSocketFactory: 	at com.fsck.k9.mail.ssl.DefaultTrustedSocketFactory.setSniHost(DefaultTrustedSocketFactory.java:152)
DefaultTrustedSocketFactory: 	at com.fsck.k9.mail.ssl.DefaultTrustedSocketFactory.createSocket(DefaultTrustedSocketFactory.java:133)
DefaultTrustedSocketFactory: 	at com.fsck.k9.mail.transport.smtp.SmtpTransport.open(SmtpTransport.kt:131)
DefaultTrustedSocketFactory: 	at com.fsck.k9.mail.transport.smtp.SmtpTransport.sendMessage(SmtpTransport.kt:349)
DefaultTrustedSocketFactory: 	at com.fsck.k9.backend.imap.ImapBackend.sendMessage(ImapBackend.kt:153)
DefaultTrustedSocketFactory: 	at com.fsck.k9.controller.MessagingController.sendPendingMessagesSynchronous(MessagingController.java:1559)
DefaultTrustedSocketFactory: 	at com.fsck.k9.controller.MessagingController$7.run(MessagingController.java:1449)
DefaultTrustedSocketFactory: 	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:222)
DefaultTrustedSocketFactory: 	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:108)
DefaultTrustedSocketFactory: 	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:163)
DefaultTrustedSocketFactory: 	at java.lang.Thread.run(Thread.java:923)
SmtpTransport: SMTP >>> EHLO [127.0.0.1]
SmtpTransport: SMTP >>> QUIT
MessagingController: Failed to send message
MessagingController: com.fsck.k9.mail.MessagingException: Unable to open connection to SMTP server.
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpTransport.open(SmtpTransport.kt:251)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpTransport.sendMessage(SmtpTransport.kt:349)
MessagingController: 	at com.fsck.k9.backend.imap.ImapBackend.sendMessage(ImapBackend.kt:153)
MessagingController: 	at com.fsck.k9.controller.MessagingController.sendPendingMessagesSynchronous(MessagingController.java:1559)
MessagingController: 	at com.fsck.k9.controller.MessagingController$7.run(MessagingController.java:1449)
MessagingController: 	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:222)
MessagingController: 	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:108)
MessagingController: 	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:163)
MessagingController: 	at java.lang.Thread.run(Thread.java:923)
MessagingController: Caused by: java.net.SocketException: Connection reset
MessagingController: 	at java.net.SocketInputStream.read(SocketInputStream.java:215)
MessagingController: 	at java.net.SocketInputStream.read(SocketInputStream.java:144)
MessagingController: 	at java.io.BufferedInputStream.fill(BufferedInputStream.java:248)
MessagingController: 	at java.io.BufferedInputStream.read(BufferedInputStream.java:267)
MessagingController: 	at com.fsck.k9.mail.filter.PeekableInputStream.read(PeekableInputStream.java:27)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpResponseParser.readByte(SmtpResponseParser.kt:266)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpResponseParser.readChar(SmtpResponseParser.kt:274)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpResponseParser.readDigit(SmtpResponseParser.kt:254)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpResponseParser.readReplyCode1(SmtpResponseParser.kt:234)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpResponseParser.readReplyCode(SmtpResponseParser.kt:230)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpResponseParser.readHelloResponse(SmtpResponseParser.kt:36)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpTransport.sendHello(SmtpTransport.kt:306)
MessagingController: 	at com.fsck.k9.mail.transport.smtp.SmtpTransport.open(SmtpTransport.kt:142)
MessagingController: 	... 8 more
LockableDatabase: LockableDatabase: Transaction ended, took 11 ms / com.fsck.k9.mailstore.LocalMessage.setFlag(LocalMessage.java:260)
LockableDatabase: LockableDatabase: Transaction ended, took 14 ms / com.fsck.k9.mailstore.LocalFolder.destroyMessage(LocalFolder.java:1002)
LockableDatabase: LockableDatabase: Transaction ended, took 13 ms / com.fsck.k9.mailstore.LocalFolder.destroyMessage(LocalFolder.java:1002)
MessagingController:  Command 'sendPendingMessages' completed
```

--- a/mail/common/src/main/java/com/fsck/k9/mail/Transport.java
+++ b/mail/common/src/main/java/com/fsck/k9/mail/Transport.java
@@ -2,12 +2,6 @@
 package com.fsck.k9.mail;
 
 public abstract class Transport {
-
-    protected static final int SOCKET_CONNECT_TIMEOUT = 10000;
-
-    // RFC 1047
-    protected static final int SOCKET_READ_TIMEOUT = 300000;
-
     public abstract void open() throws MessagingException;
 
     public abstract void sendMessage(Message message) throws MessagingException;

--- a/mail/protocols/smtp/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.kt
+++ b/mail/protocols/smtp/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.kt
@@ -130,15 +130,16 @@ class SmtpTransport(
                 if (extensions.containsKey("STARTTLS")) {
                     executeCommand("STARTTLS")
 
-                    this.socket = trustedSocketFactory.createSocket(
+                    val tlsSocket = trustedSocketFactory.createSocket(
                         socket,
                         host,
                         port,
                         clientCertificateAlias
                     )
-                    inputStream = PeekableInputStream(BufferedInputStream(socket.getInputStream(), 1024))
+                    this.socket = tlsSocket
+                    inputStream = PeekableInputStream(BufferedInputStream(tlsSocket.getInputStream(), 1024))
                     responseParser = SmtpResponseParser(logger, inputStream!!)
-                    outputStream = BufferedOutputStream(socket.getOutputStream(), 1024)
+                    outputStream = BufferedOutputStream(tlsSocket.getOutputStream(), 1024)
 
                     // Now resend the EHLO. Required by RFC2487 Sec. 5.2, and more specifically, Exim.
                     extensions = sendHello(helloName)

--- a/mail/protocols/smtp/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.kt
+++ b/mail/protocols/smtp/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.kt
@@ -10,6 +10,8 @@ import com.fsck.k9.mail.K9MailLib
 import com.fsck.k9.mail.Message
 import com.fsck.k9.mail.Message.RecipientType
 import com.fsck.k9.mail.MessagingException
+import com.fsck.k9.mail.NetworkTimeouts.SOCKET_CONNECT_TIMEOUT
+import com.fsck.k9.mail.NetworkTimeouts.SOCKET_READ_TIMEOUT
 import com.fsck.k9.mail.ServerSettings
 import com.fsck.k9.mail.Transport
 import com.fsck.k9.mail.filter.Base64

--- a/mail/protocols/smtp/src/test/java/com/fsck/k9/mail/transport/mockServer/MockSmtpServer.java
+++ b/mail/protocols/smtp/src/test/java/com/fsck/k9/mail/transport/mockServer/MockSmtpServer.java
@@ -68,6 +68,11 @@ public class MockSmtpServer {
         interactions.add(new ExpectedCommand(command));
     }
 
+    public void startTls() {
+        checkServerNotRunning();
+        interactions.add(new UpgradeToTls());
+    }
+
     public void closeConnection() {
         checkServerNotRunning();
         interactions.add(new CloseConnection());
@@ -212,6 +217,9 @@ public class MockSmtpServer {
         }
     }
 
+    private static class UpgradeToTls implements SmtpInteraction {
+    }
+
     private static class CloseConnection implements SmtpInteraction {
     }
 
@@ -303,6 +311,8 @@ public class MockSmtpServer {
                 readExpectedCommand((ExpectedCommand) interaction);
             } else if (interaction instanceof CannedResponse) {
                 writeCannedResponse((CannedResponse) interaction);
+            } else if (interaction instanceof UpgradeToTls) {
+                upgradeToTls(socket);
             } else if (interaction instanceof CloseConnection) {
                 clientSocket.close();
             }

--- a/mail/protocols/smtp/src/test/java/com/fsck/k9/mail/transport/smtp/SmtpTransportTest.kt
+++ b/mail/protocols/smtp/src/test/java/com/fsck/k9/mail/transport/smtp/SmtpTransportTest.kt
@@ -568,6 +568,35 @@ class SmtpTransportTest {
     }
 
     @Test
+    fun `open() with STARTTLS`() {
+        val server = MockSmtpServer().apply {
+            output("220 localhost Simple Mail Transfer Service Ready")
+            expect("EHLO [127.0.0.1]")
+            output("250-localhost Hello 127.0.0.1")
+            output("250-STARTTLS")
+            output("250 HELP")
+            expect("STARTTLS")
+            output("220 Ready to start TLS")
+            startTls()
+            expect("EHLO [127.0.0.1]")
+            output("250-localhost Hello 127.0.0.1")
+            output("250 AUTH PLAIN LOGIN")
+            expect("AUTH PLAIN AHVzZXIAcGFzc3dvcmQ=")
+            output("235 2.7.0 Authentication successful")
+        }
+        val transport = startServerAndCreateSmtpTransport(
+            server,
+            authenticationType = AuthType.PLAIN,
+            connectionSecurity = ConnectionSecurity.STARTTLS_REQUIRED
+        )
+
+        transport.open()
+
+        server.verifyConnectionStillOpen()
+        server.verifyInteractionCompleted()
+    }
+
+    @Test
     fun `sendMessage() without address to send to should not open connection`() {
         val message = MimeMessage()
         val server = createServerAndSetupForPlainAuthentication()

